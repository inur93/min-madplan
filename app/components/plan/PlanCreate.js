import { Form } from "../shared/Form";
import { Input } from "../shared/Input";
import { getWeek, startOfNextWeek, addDays, formatDateForQuery } from '../../functions/dateFunctions';
import { Button } from "semantic-ui-react";
import { ButtonSuccess } from "../shared/Button";

export function PlanCreate({ onCreate, loading }) {
    const nextWeek = startOfNextWeek();
    const week = getWeek(nextWeek);
    const defaultDateValue = formatDateForQuery(nextWeek);
    return <Form onSubmit={onCreate}>
        <Input name="name"
            required
            defaultValue={`Uge ${week}`}
            placeholder="Giv din ugeplan et navn fx 'uge 12'" />
        <Input name="validFrom"
            required
            defaultValue={defaultDateValue} />
        <ButtonSuccess disabled={loading} fluid type='submit' name='submit' icon='success' />
    </Form>
}