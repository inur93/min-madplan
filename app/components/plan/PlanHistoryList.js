import { List } from "semantic-ui-react";
import {formatDate} from '../../functions/dateFunctions';
export function PlanHistoryList({ list, onClick }) {
    const handleClick = (id) => () => onClick(id);
    return (<List>
        {list.map(({ _id, title, validFrom }) => {            
            return <List.Item key={_id} onClick={handleClick(_id)}>
                <List.Content>
                    <List.Header as='h2'>{title}</List.Header>
                    <List as='a'>{formatDate(new Date(validFrom))}</List>
                </List.Content>
            </List.Item>
        })}
    </List>)
}