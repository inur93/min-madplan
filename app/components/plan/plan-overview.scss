.mmp-plan-list-item {
  h2 {
    font-size: medium;
  }

  .content {
    display: flex;
    flex-flow: row;
    align-items: stretch;

    
    font-size: 18px;
    .header {
      flex: auto 1 1;
      white-space: nowrap;
      overflow: hidden;
      text-overflow: ellipsis;
      margin-right: 0.5rem;
    }

    .icon {
      flex: auto 0 1;
      padding: unset;
      margin: auto 0;
      margin-left: 0.75rem;
    }
  }

  .content h2.mmp-empty.header {
    color: red;
  }
}
