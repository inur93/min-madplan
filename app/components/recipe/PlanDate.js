import { Message, Icon, List as ListSUI } from 'semantic-ui-react';
import { formatDay } from '../../functions/dateFunctions';


export function PlanDate({ loading, plan, date }) {
    if (loading) return null;
    const selected = plan.find(x => x.date === date);
    return <Message info
        header={formatDay(new Date(date))}
        content={selected ? selected.recipe.name : 'VÃ¦lg en opskrift...'} />
}