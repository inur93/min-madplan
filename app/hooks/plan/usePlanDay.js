import { GetPlanApi } from "../../_api";
import { useRouter } from "next/router";
import useSWR from "swr";
import { useState, useEffect } from "react";


export function usePlanDay(idParam, dateParam) {
    const api = GetPlanApi();
    const router = useRouter();
    const [planDay, setPlanDay] = useState(null);
    const [hide, setHide] = useState(true);
    const {
        data: plan,
        isValidating: reloading
    } = useSWR(router.query[idParam]
        ? ['/plan', router.query[idParam]]
        : null,
        (url, id) => api.findOne(id));


    useEffect(() => {
        setHide(!router.query[idParam]
            || !router.query[dateParam]);
    }, [router.query[idParam], router.query[dateParam]]);

    useEffect(() => {
        if (plan) {
            const existing = plan.plan.find(x => x.date === router.query[dateParam]);
            setPlanDay(existing || { date: router.query[dateParam] });
        }
    }, [plan, router.query[dateParam]]);

    const state = {
        hide,
        planDay,
        loading: !planDay || reloading
    }

    return [state]

}