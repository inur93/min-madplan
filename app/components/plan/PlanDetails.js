import { List, Icon } from "semantic-ui-react";
import { addDays, formatDateForQuery, formatDay, getPlanLength } from '../../functions/dateFunctions';
import { IconEdit, IconInfo, IconRemove } from "../shared/Icon";
import { usePlanDetails } from "../../hooks/plan/usePlanDetails";
import { Loader } from '../shared/Loader';
import { useState } from "react";

const ListItem = function ({ date, recipe, onEdit, onRemove, onInfo }) {
    const isEmpty = !recipe;
    const handleInfo = () => onInfo({ recipe: recipe._id, date });
    const handleRemove = async () => await onRemove({ date });
    const handleEdit = () => onEdit({ recipe: recipe && recipe._id, date });

    return (<List.Item>
        <List.Content>
            <List.Header as="a">{formatDay(new Date(date))}</List.Header>
            <List.Header onClick={handleEdit} className={isEmpty ? 'mmp-empty' : ''} >
                {isEmpty ?
                    "Tilf√∏j en opskrift..."
                    : recipe.title
                }
            </List.Header>
        </List.Content>
        <List.Content>
            <IconEdit onClick={handleEdit} />
            {!isEmpty && <IconInfo onClick={handleInfo} />}
            {!isEmpty && <IconRemove onClick={handleRemove} />}
        </List.Content>
    </List.Item>)
}

export function PlanDetails() {
    const [state, handlers] = usePlanDetails();

    const onEdit = (data) => handlers.editPlanDay({ ...data, id: state.plan._id });
    const onInfo = (data) => handlers.infoPlanDay({ ...data, id: state.plan._id });

    return (<Loader loading={state.loading}>
        <h2>{state.plan && state.plan.name}</h2>
        <List>
            {state.planDays.map(({ date, recipe }) => {
                return <ListItem key={date}
                    onEdit={onEdit}
                    onRemove={handlers.removePlanDay}
                    onInfo={onInfo}
                    recipe={recipe}
                    date={date} />
            })}
        </List>
    </Loader>)
}