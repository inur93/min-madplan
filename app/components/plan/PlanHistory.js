import { List, Icon } from "semantic-ui-react";
import { formatDate } from '../../functions/dateFunctions';
import { useView } from "../../hooks/shared/useView";
import { Loader } from "../shared/Loader";
import { usePlanHistory } from "../../hooks/plan/usePlanHistory";

const EmptyMessage = () => <p>
    Du har ikke nogen madplan for den kommende uge, 
    opret en ved at trykke p√• {<Icon name='calendar plus outline' />}.
    </p>

export function PlanHistory() {
    const [state, handlers] = usePlanHistory();
    const [show, edit, goTo] = useView('/plan');
    return (<Loader loading={state.loading}>
        <List>
            {!state.history.length && <EmptyMessage />}
            {state.history.map(({ _id, name, validFrom }) => {
                return <List.Item key={_id} onClick={() => goTo.details(_id)}>
                    <List.Content>
                        <List.Header as='h2'>{name}</List.Header>
                        <List.Description as='a'>{formatDate(new Date(validFrom))}</List.Description>
                    </List.Content>
                </List.Item>
            })}
        </List>
    </Loader>)
}