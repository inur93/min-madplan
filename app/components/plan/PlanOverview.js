import { List } from "semantic-ui-react";
import { getPlanLength, addDays, formatDateForQuery, formatDay } from '../../functions/dateFunctions';
import './plan-overview.scss';

const ListItem = function ({ date, recipe, onClick }) {
    const isEmpty = !recipe;
    return (<List.Item className="mmp-plan-list-item" onClick={onClick}>
        <List.Content>
            <List.Header as="a">{formatDay(date)}</List.Header>
            <List.Header className={isEmpty ? 'mmp-empty' : ''} as="h2" >
                {isEmpty ?
                    "Tilf√∏j en opskrift..."
                    : recipe.title
                }
            </List.Header>
        </List.Content>
    </List.Item>)
}

export function PlanView({ plan, onClick }) {

    const { _id, validFrom, length, durationType } = plan;
    if (!validFrom) return <p>empty...</p>;
    let days = new Array(getPlanLength(length, durationType)).fill(null);
    let baseDate = new Date(validFrom);
    const handleClick = (date) => () => onClick(_id, date);
    return (<List>
        {days.map((val, i) => {
            const date = addDays(baseDate, i);
            const queryDate = formatDateForQuery(date);
            const entry = plan.plan.find(p => p.date === queryDate);
            return <ListItem key={date}
                onClick={handleClick(queryDate)}
                recipe={entry && entry.recipe}
                date={date} />
        })}
    </List>)
}