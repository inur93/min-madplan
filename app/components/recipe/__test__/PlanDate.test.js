import React from "react";
import { render, unmountComponentAtNode } from "react-dom";
import { PlanDate } from "../PlanDate";
import renderer from "react-test-renderer";
import { useRecipeDetails, usePlanDay } from '../../../hooks/plan/usePlanDay';
jest.mock("../../../hooks/plan/usePlanDay");

let container = null;
beforeEach(() => {
    container = document.createElement("div");
    document.body.appendChild(container);
})

afterEach(() => {
    unmountComponentAtNode(container);
    container.remove();
    container = null;
})

describe("Plan date view on recipe", () => {
    it("no plan date selected", () => {
        usePlanDay.mockReturnValue([{ hide: true }])
        const tree = renderer
            .create(<PlanDate />)
            .toJSON();
        expect(tree).toMatchSnapshot();
    })
    it("loading while empty", () => {
        usePlanDay.mockReturnValue([{ loading: true }])
    })

    it("showing recipe", () => {
        usePlanDay.mockReturnValue([{
            recipe: {
                title: 'En rigtig fin opskrift',
                date: '2020-04-25'
            }
        }])
    })
})
