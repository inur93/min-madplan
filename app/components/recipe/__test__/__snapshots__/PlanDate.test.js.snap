// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Plan date view on recipe no plan date selected 1`] = `null`;
